libraries/push/impl/src/main/kotlin/io/element/android/libraries/push/impl/notifications/NotificationResolverQueue.kt

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2025 New Vector Ltd.
+ *
+ * SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
+ * Please see LICENSE files in the repository root for full details.
+ */
+
+package io.element.android.libraries.push.impl.notifications
+
+import io.element.android.libraries.di.AppScope
+import io.element.android.libraries.di.SingleIn
+import io.element.android.libraries.matrix.api.core.EventId
+import io.element.android.libraries.matrix.api.core.RoomId
+import io.element.android.libraries.matrix.api.core.SessionId
+import io.element.android.libraries.push.impl.notifications.model.ResolvedPushEvent
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.SharedFlow
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.launch
+import timber.log.Timber
+import javax.inject.Inject
+import kotlin.time.Duration.Companion.milliseconds
+
+/**
+ * This class is responsible for periodically batching notification requests and resolving them in a single call,
+ * so that we can avoid having to resolve each notification individually in the SDK.
+ */
+@OptIn(ExperimentalCoroutinesApi::class)
+@SingleIn(AppScope::class)
+class NotificationResolverQueue @Inject constructor(
+    private val notifiableEventResolver: NotifiableEventResolver,
+    private val coroutineScope: CoroutineScope,
+) {
+    companion object {
+        private const val BATCH_WINDOW_MS = 250L
+    }
+    private val requestQueue = Channel<NotificationEventRequest>(capacity = 100)
+
+    /**
+     * A flow that emits pairs of a list of notification event requests and a map of the resolved events.
+     * The map contains the original request as the key and the resolved event as the value.
+     */
+    val results: SharedFlow<Pair<List<NotificationEventRequest>, Map<NotificationEventRequest, Result<ResolvedPushEvent>>>> = MutableSharedFlow()
+
+    init {
+        coroutineScope.launch {
+            while (coroutineScope.isActive) {
+                // Wait for a batch of requests to be received in a specified time window
+                delay(BATCH_WINDOW_MS.milliseconds)
```

Comment:
Are you sure this code will behave as a first throttler?
If we get a Push just before the delay is ended, it will trigger the request without waiting for any other potential push. I have not checked in practice how much time we have between mutliple pending push received from Firebase though.